tket/src/subcircuit.rs

```diff
@@ -0,0 +1,167 @@
+//! Subcircuits of circuits.
+//!
+//! This provides a specialised version of hugr's [`SiblingSubgraph`], for the
+//! case where the HUGR region is a [`Circuit`].
+
+use hugr::hugr::views::sibling_subgraph::{InvalidReplacement, InvalidSubgraph};
+use hugr::hugr::views::SiblingSubgraph;
+use hugr::HugrView;
+use itertools::Itertools;
+use portgraph::algorithms::convex::{LineIndex, LineInterval, LineIntervals};
+
+use crate::circuit::Circuit;
+use crate::rewrite::CircuitRewrite;
+
+pub use hugr::hugr::views::sibling_subgraph::LineConvexChecker;
+
+/// A subcircuit of a circuit.
+///
+/// Store convex subgraphs of Circuit-like regions of HUGRs as intervals on the
+/// "lines" of the circuit (see below and [`LineConvexChecker`]).
+///
+/// Lines are the result of a partition of the DFG
+/// into edge-disjoint paths; in the case of purely quantum ops, the lines are
+/// the qubits
+///
+/// ## Differences with [`SiblingSubgraph`]
+///
+/// [`Subcircuit`]s have two main distinguishing features over
+/// [`SiblingSubgraph`], that make it more suitable in some contexts, especially
+/// when working with subgraphs of circuits:
+///  - The representation of the subcircuit uses the circuit structure to be more
+///    memory efficient. The size of a [`Subcircuit`] instance is not linear in
+///    the number of nodes in the subgraph, but linear in the number of lines
+///    (~qubits) in the subcircuit.
+///  - Subcircuits can be updated by extending the intervals without having to
+///    recompute the subgraph from scratch and rechecking convexity.
+///
+/// Along with these features come two caveats to be aware of:
+///  - Unlike subgraphs, [`Subcircuit`]s do not maintain an ordered boundary.
+///    This is because the ordering would be hard to maintain when the subcircuit
+///    is expanded. When a subcircuit is converted into a [`SiblingSubgraph`],
+///    a boundary order gets fixed.
+///  - Constructing a [`LineConvexChecker`] (required to construct [`Subcircuit`]s)
+///    is currently ~2.5x more expensive than constructing the default
+///    [`TopoConvexChecker`]. This must be amortised over many subcircuit
```

Comment:
```suggestion
///    [`TopoConvexChecker`]. This may be amortised over many subcircuit
```
(the caller does that by reusing the `LineConvexChecker` instance)